Add Elytron ldap cache

jonasjensen-brolstark opened the pull request:
Add configuration for Elytron ldap cache.

- Closes: #10267

JonasJensen committed:
add elytron LDAP cache

sberyozkin reviewed (approved):
LGTM, @michalvavrik, do you agree ?

michalvavrik commented on extensions/elytron-security-ldap/runtime/src/main/java/io/quarkus/elytron/security/ldap/config/CacheConfig.java:
```diff
@@ -0,0 +1,30 @@
+public interface CacheConfig {
+
+    /**
+     * If set to true, request to the ldap server are cached
```
```suggestion
     * If set to true, request to the LDAP server are cached
```

michalvavrik commented on extensions/elytron-security-ldap/runtime/src/main/java/io/quarkus/elytron/security/ldap/LdapRecorder.java:
```diff
@@ -47,7 +50,14 @@ public RuntimeValue<SecurityRealm> createRealm(LdapSecurityRealmRuntimeConfig ru
-        return new RuntimeValue<>(ldapSecurityRealmBuilder.build());
+        SecurityRealm ldapRealm = ldapSecurityRealmBuilder.build();
+
+        if (ldapRealm instanceof CacheableSecurityRealm && runtimeConfig.cache().enabled()) {
```
`ldapSecurityRealmBuilder` currently builds `LdapSecurityRealm` which `CacheableSecurityRealm`, but if that should ever change, we will ignore enabled cache and user will not know. That should be illegal state, or some warning etc.

jonasjensen-brolstark replied:
Fail point. I will add a warning log to the user that the configuration will be ignored.

michalvavrik commented on extensions/elytron-security-ldap/deployment/src/test/java/io/quarkus/elytron/security/ldap/CacheTest.java:
```diff
@@ -0,0 +1,30 @@
+import io.restassured.RestAssured;
+
+@QuarkusTestResource(LdapServerTestResource.class)
+public class CacheTest {
```
Problem with this test is that it would pass even on the current main where there is no caching feature. I think proper test would require stopping the `LdapServerTestResource`. That however, would require to implement `io.quarkus.test.common.QuarkusTestResourceLifecycleManager#inject(io.quarkus.test.common.QuarkusTestResourceLifecycleManager.TestInjector)`, injecting it into this test, stop the LDAP server and make request again.

How does it sound? As Sergey already approved it is not required ...

sberyozkin replied:
@michalvavrik It is required because you noticed it :-). I approved as I thought the actual code change done by @jonasjensen-brolstark was good and simple, but I was delaying  enabling the CI run since your approval is also required for this PR. Thanks for the review so far

jonasjensen-brolstark replied:
Good catch. I will try to give it a try and implement that test.

michalvavrik replied:
Thank you!

jonasjensen-brolstark commented:
@michalvavrik i have added your suggestions and added a IT test, where the LDAP server i shutdown between requests.

I am bit in doubt about the warning text. So if you have any comments/suggestions it would be nice 🙂 

I will squash my commits, if you approve.

sberyozkin commented:
Thanks @jonasjensen-brolstark, @michalvavrik, Michal, if you are happy, please add your approval too

michalvavrik reviewed (commented):
Sorry @sberyozkin about the review, I remembered to request your review, but forgot to submit my comments that I had here.

michalvavrik commented:
Anyway, I think current code is good, thanks @jonasjensen-brolstark .

sberyozkin commented on extensions/elytron-security-ldap/runtime/src/main/java/io/quarkus/elytron/security/ldap/LdapRecorder.java:
```diff
@@ -47,7 +54,18 @@ public RuntimeValue<SecurityRealm> createRealm(LdapSecurityRealmRuntimeConfig ru
+                ldapRealm = new CachingSecurityRealm(ldapRealm,
+                        new LRURealmIdentityCache(runtimeConfig.cache().size(), runtimeConfig.cache().maxAge().toMillis()));
+            } else {
+                log.warn("LDAP realm not cacheable, ignoring 'cache' configuration");
```
@michalvavrik I'm definitely not the best doc reviewer :-). Your suggestion is fine but I'm not sure if users know what `SecurityRealm` is. The original proposed warning looks OK to me, with tiny updates, let me suggest

sberyozkin commented on extensions/elytron-security-ldap/runtime/src/main/java/io/quarkus/elytron/security/ldap/LdapRecorder.java:
```diff
@@ -47,7 +54,18 @@ public RuntimeValue<SecurityRealm> createRealm(LdapSecurityRealmRuntimeConfig ru
+                ldapRealm = new CachingSecurityRealm(ldapRealm,
+                        new LRURealmIdentityCache(runtimeConfig.cache().size(), runtimeConfig.cache().maxAge().toMillis()));
+            } else {
+                log.warn("LDAP realm not cacheable, ignoring 'cache' configuration");
```
```suggestion
                log.warn("LDAP realm is not cacheable, ignoring the 'cache' configuration");
```

sberyozkin replied:
@michalvavrik Is it reading all right ?

michalvavrik replied:
sure, thanks

sberyozkin replied:
@jonasjensen-brolstark Hi, please choose one of the proposed doc updates, either from Michal or the one I just suggested, and then we'll get a build running and the PR merged, thanks

jonasjensen-brolstark replied:
is done 🙂

sberyozkin commented on test-framework/ldap/src/main/java/io/quarkus/test/ldap/LdapServerTestResource.java:
```diff
@@ -49,4 +49,9 @@ public synchronized void stop() {
+
+    @Override
+    public void inject(TestInjector testInjector) {
+        testInjector.injectIntoFields(ldapServer, new TestInjector.MatchesType(InMemoryDirectoryServer.class));
```
If the test is good and works, it is fine I guess

michalvavrik commented on integration-tests/elytron-security-ldap/src/test/java/io/quarkus/elytron/security/ldap/it/ElytronSecurityLdapTest.java:
```diff
@@ -82,4 +87,26 @@ void admin_role_not_authorized() {
+                .then()
+                .statusCode(200);
+
+        ldapServer.shutDown(false);
```
this uses default test method order, so if you stop this server here, other test method results can be corrupted (yet succeed?). wdyt can you make this run as the last, please?

jonasjensen-brolstark replied:
You are right. It is pure luck that this test runs last.

So i can either user the `@Order` annotation on all the tests in the class or simply start the server again in the end of the test.

@michalvavrik do you have any preferences?

michalvavrik replied:
both is fine with me as it is in memory handler, it should be quick. personally I'd go with `@Order` but you are the PR author!

jonasjensen-brolstark replied:
Is done.

michalvavrik reviewed (approved):
I am fine with changes, but I think the place where you put the test with shutting down server is suspicious. I'd make it declaratively last as the default method order depends on complex algorithm and I'm not able to say if it is last. anyway, good job, thanks

sberyozkin commented:
@jonasjensen-brolstark FYI:

```
Caused by: org.apache.maven.plugin.MojoFailureException: File '/home/runner/work/quarkus/quarkus/extensions/elytron-security-ldap/runtime/src/main/java/io/quarkus/elytron/security/ldap/LdapRecorder.java' has not been previously formatted. Please format file (for example by invoking `mvn -f extensions/elytron-security-ldap/runtime net.revelc.code.formatter:formatter-maven-plugin:2.23.0:format`) and commit before running validation!
```

geoand commented:
I updated and force pushed the change

sberyozkin commented:
Thanks @geoand

geoand commented:
👍🏼
